File: apps/api/routers/v1/schemas/ai_chat.py
# -*- coding: utf-8 -*-
"""Схемы веб-чата с ассистентом (POST /ai/chat)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class ChatMessageIn(BaseModel):
    """Одно сообщение истории (историю держит клиент и шлёт с каждым запросом)."""

    role: Literal["user", "assistant"]
    content: str = Field(min_length=1, max_length=4000)


class AIChatRequest(BaseModel):
    """Запрос чата: история (последнее сообщение — вопрос пользователя)."""

    messages: list[ChatMessageIn] = Field(min_length=1, max_length=12)


class ToolCallOut(BaseModel):
    """След вызова инструмента — фронт показывает «что ассистент проверял»."""

    name: str
    error: str | None = None


class AIChatResponse(BaseModel):
    """Ответ ассистента."""

    answer: str
    tool_calls: list[ToolCallOut] = Field(default_factory=list)
    generated_at: str
    model: str


class AIPulseResponse(BaseModel):
    """Почасовой пульс для веб-виджета.

    important=False → за окно ничего значимого, виджет молчит (text = null).
    """

    important: bool
    text: str | None = None
    generated_at: str

File: frontend/src/lib/api/aiChat.ts
/**
 * API-модуль веб-чата с AI-ассистентом (плавающий виджет дашборда).
 *
 * Эндпоинты:
 *   POST /api/ai/chat   →  { answer, tool_calls, generated_at, model }
 *   POST /api/ai/pulse  →  { important, text, generated_at }
 *
 * Историю диалога держит КЛИЕНТ (стор chatWidget) — сервер её не хранит.
 * В запрос уходят последние ≤12 сообщений (role: user|assistant, content ≤4000
 * символов), последним — вопрос пользователя.
 *
 * Пульс: сервер кэширует результат на календарный час (UTC) — виджет может
 * опрашивать смело, AI дёргается максимум раз в час глобально. important=false
 * (text=null) — за час ничего значимого, виджет молчит.
 *
 * Ошибки: 429 (лимит 30/час), 503 (AI-провайдеры не настроены) — приходят как
 * ApiError через apiSend, разбираются в сторе.
 */
import type { components } from "@fb/shared/api/generated";
import { ApiError } from "./client";
import { generatedFetchApi } from "./generatedClient";

export type AiChatRole = components["schemas"]["ChatMessageIn"]["role"];
export type AiChatMessageIn = components["schemas"]["ChatMessageIn"];
export type AiChatResponse = components["schemas"]["AIChatResponse"];
export type AiPulseResponse = components["schemas"]["AIPulseResponse"];
export type AiChatToolCall = NonNullable<AiChatResponse["tool_calls"]>[number];

function generatedApiError(status: number, error: unknown): ApiError {
  const message =
    typeof error === "object" &&
    error !== null &&
    "message" in error &&
    typeof error.message === "string"
      ? error.message
      : `Ошибка API ${status}`;
  return new ApiError(message, status, error);
}

/** Отправляет историю диалога и возвращает ответ ассистента. */
export async function sendAiChatMessage(messages: AiChatMessageIn[]): Promise<AiChatResponse> {
  const { data, error, response } = await generatedFetchApi.POST("/api/ai/chat", { body: { messages } });
  if (!response.ok) throw generatedApiError(response.status, error);
  if (!data) throw new Error(`Пустой ответ API: ${response.status}`);
  return data;
}

/** Почасовой пульс кабинета: important=false → тишина, виджет молчит. */
export async function fetchAiPulse(): Promise<AiPulseResponse> {
  const { data, error, response } = await generatedFetchApi.POST("/api/ai/pulse");
  if (!response.ok) throw generatedApiError(response.status, error);
  if (!data) throw new Error(`Пустой ответ API: ${response.status}`);
  return data;
}
